src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java

```diff
@@ -417,14 +419,26 @@ public long countAttributesByGUID(final String gemrplasmUUID, final List<String>
 		return this.daoFactory.getAttributeDAO().countAttributesByGUID(gemrplasmUUID, attributeDbIds);
 	}
 
-  private List<Integer> getGidsOfGermplasmWithDerivativeOrMaintenanceDescendants(final List<Germplasm> germplasmList) {
-	  final List<Integer> gids = germplasmList.stream().map(Germplasm::getGid).collect(Collectors.toList());
+	@Override
+	public Germplasm getGermplasmWithPreferredName(final Integer gid) {
+		final Germplasm germplasm = this.daoFactory.getGermplasmDao().getById(gid);
+		if (germplasm != null) {
+			// Old data has germplasm with multiple preferred names, get the first preferred name for those germplasm
+			// Replace with germplasm.setPreferredName(this.daoFactory.getNameDao().getPreferredNameByGid(gid)) when IBP-4596 is resolved
+			final List<Name> names = this.daoFactory.getNameDao().getByGIDWithFilters(gid, 1, null);
+			if (!names.isEmpty()) {
+				germplasm.setPreferredName(names.get(0));
+			}
+		}
+		return germplasm;
+	}
 
-	  // Get all DER/MAN germplasm that has existing derivative progeny.
-		return Lists.newArrayList(this.daoFactory.getGermplasmDao().getGidsOfGermplasmWithDerivativeOrMaintenanceDescendants(Sets.newHashSet(gids)));
-  }
+	@Override
+	public List<Germplasm> getProgenitorsWithPreferredName(final Integer gid) {
+		return this.daoFactory.getGermplasmDao().getProgenitorsByGIDWithPrefName(gid);
```

Comment:
Again, we should not be returning Germplasm class, if possible. 
Further, the existing GermplasmDAO#getProgenitorsByGIDWithPrefName is not good: it is querying from `listdata` for the `malePedigree` field although seems that is not needed in this context. 
Check if we can return `GermplasmDto` instead

Reply:
Also I don't like naming methods : `getXXXWithPreferredName`. 
The DTO returned by the method should tell us what information is available (let's make sure it's available as much as possible)

Reply:
Done